examples/secret_storage/secret_storage.go

```diff
@@ -0,0 +1,572 @@
+something similar to hide namespaces.
+
+Additionally this example handles concurrent writes by using explicit revision
+numbers to prevent one user from unintentionally clobbering another user's
+rental updates.
+
+*/
+package main
+
+import (
+	"crypto/hmac"
+	"crypto/rand"
+	"crypto/sha256"
+	"encoding/hex"
+	"encoding/json"
+	"flag"
+	"fmt"
+	"log"
+	"os"
+	"strings"
+
+	"github.com/keybase/go-keybase-chat-bot/kbchat"
+	"github.com/keybase/go-keybase-chat-bot/kbchat/types/keybase1"
+
+	"golang.org/x/sync/errgroup"
+)
+
+// SecretKeyKVStoreAPI is a KVStoreAPI that hides the entryKeys from Keybase servers.
+// It does so by HMACing entryKeys using a per-(team, namespace) secret,
+// and storing the HMAC instead of the plaintext entryKey. This approach
+// does not handle any secret rotation, and does not expect the secret to
+// change.
+//
+// The plaintext entryKey is stored in it's corresponding JSON entryValue
+// under the key "_key" to enable listing.
+//
+// This approach does not hide memory access patterns. Also, Keybase
+// servers prevent a removed team member from continuing to access a team's
+// data, but if that were somehow bypassed*, a former team member who still
+// knows the HMAC secret can check for the presence of specific entryKeys
+// (*but you probably have bigger issues to deal with in that case...).
+type SecretKeyKVStoreAPI struct {
+	api       kbchat.KVStoreAPI
+	secrets   map[string](map[string][]byte)
+	secretKey string
+	keyKey    string
+}
+
+func NewSecretKeyKVStoreAPI(api kbchat.KVStoreAPI) *SecretKeyKVStoreAPI {
+	secrets := make(map[string](map[string][]byte))
+	sc := SecretKeyKVStoreAPI{api, secrets, "_secret", "_key"}
+	return &sc
+}
+
+func (sc *SecretKeyKVStoreAPI) loadSecret(teamName string, namespace string) ([]byte, error) {
+	if _, ok := sc.secrets[teamName]; !ok {
+		sc.secrets[teamName] = make(map[string][]byte)
+	}
+	if secret, ok := sc.secrets[teamName][namespace]; ok {
+		return secret, nil
+	}
+
+	newSecret := make([]byte, sha256.BlockSize)
+	if _, err := rand.Read(newSecret); err != nil {
+		return nil, err
+	}
+
+	// we don't expect SecretKey's revision > 0
+	_, err := sc.api.PutEntryWithRevision(teamName, namespace, sc.secretKey, hex.EncodeToString(newSecret), 1)
+	if err != nil {
+		if e, ok := err.(kbchat.Error); !ok || e.Code != kbchat.RevisionErrorCode {
+			// unexpected error
+			return nil, err
+		}
+
+		// failed to put; get entry
+		res, err := sc.api.GetEntry(teamName, namespace, sc.secretKey)
+		if err != nil {
+			return nil, err
+		}
+		existingSecret, err := hex.DecodeString(res.EntryValue)
+		if err != nil {
+			return nil, err
+		}
+		sc.secrets[teamName][namespace] = existingSecret
+	} else {
+		sc.secrets[teamName][namespace] = newSecret
+	}
+	return sc.secrets[teamName][namespace], nil
+}
+
+func (sc *SecretKeyKVStoreAPI) hmacKey(teamName string, namespace string, entryKey string) (string, error) {
```

Comment:
i left this comment on the python one too, but i still feel this way so i'm going to write it again. i don't like the name of this method. given how much it's doing and when/where it's called, i'd argue it should be called something like `obfuscateEntryKey`. calling it hmacKey makes me think that its only doing an hmac (it's not, it's also loading the secret) and that it might be called from multiple places that all require hmac functionality (it isn't because all it's doing is obfuscating an entry key). this is also the kind of thing that, in a production system, is likely to change (i.e. now we're going to hmac from a hash). at that point, the method name also needs to change. this is a big comment for a small thing, but poor naming choices can really add up to make code harder to read.

Reply:
renamed :+1: